snapcraft/remote/git.py

```diff
@@ -0,0 +1,190 @@
+        try:
+            self._repo.create_commit("HEAD", author, author, message, tree, target)
+        except pygit2.GitError as error:
+            raise GitError(
+                "Could not create a commit for the git repository "
+                f"in {str(self.path)!r}."
+            ) from error
+
+    def is_clean(self) -> bool:
+        """Check if the repo is clean.
+
+        :returns: True if the repo is clean.
+
+        :raises GitError: if git fails while checking if the repo is clean
+        """
+        try:
+            # for a clean repo, `status()` will return an empty dict
+            return not bool(self._repo.status())
+        except pygit2.GitError as error:
+            raise GitError(
+                f"Could not check if the git repository in {str(self.path)!r} is clean."
+            ) from error
+
+    def _init_repo(self) -> None:
+        """Initialize a git repo.
+
+        :raises GitError: if the repo cannot be initialized
+        """
+        logger.debug("Initializing git repository in {str(self.path)!r}")
+
+        try:
+            pygit2.init_repository(self.path)
+        except pygit2.GitError as error:
+            raise GitError(
+                f"Could not initialize a git repository in {str(self.path)!r}."
+            ) from error
+
+    def push_url(self, remote_url: str, remote_branch: str, ref: str = "HEAD") -> None:
+        """Push a reference to a branch on a remote url.
+
+        :param remote_url: the remote repo URL to push to
+        :param remote_branch: the branch on the remote to push to
+        :param ref: name of shorthand ref to push (i.e. a branch, tag, or `HEAD`)
+
+        :raises GitError: if the ref cannot be resolved or pushed
+        """
+        resolved_ref = self._resolve_ref(ref)
+        refspec = f"{resolved_ref}:refs/heads/{remote_branch}"
+
+        logger.debug(
+            "Pushing %r to remote %r with refspec %r.", ref, remote_url, refspec
+        )
+
+        try:
+            self._repo.remotes.create_anonymous(remote_url).push([refspec])
+        except pygit2.GitError as error:
+            raise GitError(
+                f"Could not push {ref!r} to {remote_url!r} with refspec {refspec!r} "
+                f"for the git repository in {str(self.path)!r}."
+            ) from error
+
+    def _resolve_ref(self, ref: str) -> str:
+        """Get a full reference name for a shorthand ref.
+
+        :param ref: shorthand ref name (i.e. a branch, tag, or `HEAD`)
+
+        :returns: the full ref name (i.e. `refs/heads/main`)
+
+        raises GitError: if the name could not be resolved
+        """
+        try:
+            reference = self._repo.lookup_reference_dwim(ref).name
+            logger.debug("Resolved reference %r for name %r", reference, ref)
+            return reference
+        # raises a KeyError if the ref does not exist and a GitError for git errors
+        except (pygit2.GitError, KeyError) as error:
+            raise GitError(
```

Comment:
this could be a different error inheriting from GitError if you plan to treat it differently

Reply:
I don't believe I will handle this differently so I'll leave it as-is for now.